### PR TITLE
[Slay the Spire] Enable support for modded characters, and add downfall support

### PR DESCRIPTION
**Pull request by KonoTyran**
## What is this fixing or adding?
changes character option to be 
`TextChoice` so custom character strings can be added

## How was this tested?

localy, via webhost.py, generage, and multiserver

### CODE REVIEW

**Comment by Alchav** [comment]
If someone spells a character name wrong, is it going to generate an unplayable game?

**Comment by KonoTyran** [comment]
no, it will default back to The Ironclad if the given name does not exist in its list of custom characters. this check is done client side.

**Comment by Berserker66** [comment]
This might be a good opportunity to also bump up the required client version, so that only a client that can use this option is allowed to connect.

**Comment by KonoTyran** [comment]
there are already incompatible clients out there using the same protocol version, that is set by the java library, and not controllable by the mod. the `client version = protocol version` breaks down when you use external libraries and clients, instead of ones built into AP directly.

**Comment by Berserker66** [comment]
"breaks down when you use external libraries and clients, instead of ones built into AP directly."
To my knowledge, the java lib is then the only one where this is hard-linked.

File: worlds/spire/__init__.py [hunk 1]
```diff
@@ -32,14 +32,15 @@ class SpireWorld(World):
...
+    required_client_version = (0, 3, 7)
 
     item_name_to_id = {name: data.code for name, data in item_table.items()}
     location_name_to_id = location_table
 
     def _get_slot_data(self):
         return {
             'seed': "".join(self.multiworld.slot_seeds[self.player].choice(string.ascii_letters) for i in range(16)),
-            'character': self.multiworld.character[self.player],
+            'character': self.multiworld.character[self.player].get_option_name,
```
Comment (Berserker66) [inline worlds/spire/__init__.py:43]: get_option_name is a method that needs to be called, as it stands you're embedding the method itself, which will fail transfer to client, as python method bytecode is not json serializable.

**Comment by Jarno458** [comment]
> "breaks down when you use external libraries and clients, instead of ones built into AP directly."
> To my knowledge, the java lib is then the only one where this is hard-linked.

To some extend the same is true for the. Net lib, sure a client can override the version that it sends. But ultimately the library knows what is supports or not, a client can claim any version. Doesn't mean the library actually supports that version

So it should ideally be up to the lib to send the version not the client, especially when the lib abstracts away the AP protocol

File: worlds/spire/__init__.py (hunk 1 above)
Comment (KonoTyran) [reply worlds/spire/__init__.py:43]: huh.... thats odd because this does work and properly sends the character string.

Comment (Berserker66) [reply worlds/spire/__init__.py:43]: I printed it to double check and it was the method, at least for me.

Comment (alwaysintreble) [reply worlds/spire/__init__.py:43]: While this works `current_key` is slightly better since it's a property

Comment (Berserker66) [reply worlds/spire/__init__.py:43]: oh yeah.. at new line 80 it gets overwritten with option.value, shadowing that this happened at all. Should probably be fixed then.

Comment (KonoTyran) [reply worlds/spire/__init__.py]: yea that was silly. fixed.

Comment (alwaysintreble) [reply worlds/spire/__init__.py]: this will still return an integer for the choices from the character selection... does the client parse those correctly?

Comment (KonoTyran) [reply worlds/spire/__init__.py]: Yea the client converts this to a string then looks for the string of "1", "2", "3", or "4" and maps them to the appropriate character/random option. Failing this it then uses that string to search all titles of installed characters. Failing that it says you get ironclad.

File: worlds/spire/Options.py [hunk 2]
```diff
@@ -37,5 +59,6 @@ class HeartRun(Toggle):
 spire_options: typing.Dict[str, type(Option)] = {
     "character": Character,
     "ascension": Ascension,
-    "heart_run": HeartRun
+    "final_act": FinalAct,
```
Comment (KonoTyran) [inline worlds/spire/Options.py:62]: yes, my mod will accept the old name of this as well for backwards compatibility.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**